### ISSUE TITLE
Stop using Django's default DB migration system

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#1106]
Sigh, this is a sad one, but it looks like Django's DB migration system is just...not good enough. Currently we use it on our master DB and then use SQL on all our replicas, so this ticket is about changing the process to using SQL everywhere. 

Some of the problems:

1. When adding text columns (which per convention are not nullable), Django's migrations use `ALTER` statements that are just...brutally bad. Like, everybody knows they're bad and yet it uses them anyway.

    Here's an example from a recent migration:

        ALTER TABLE "search_opinion" ADD COLUMN "joined_by_str" text DEFAULT '' NOT NULL;
        ALTER TABLE "search_opinion" ALTER COLUMN "joined_by_str" DROP DEFAULT;

    This migration adds a new column, sets the default value to blank, and makes it not nullable. Then in a second `ALTER`, it removes the default value. The result is that you have a column full of blanks. This is a fine outcome in Django-land, but it's terrible from a performance perspective because in Postgres < 11 (the very latest version), to create a column with a default value, you have to copy the entire table. Woo, let's hope we have disk space for that on a large table! Here's what the [postgres docs][docs] say:

    > Adding a column with a DEFAULT clause or changing the type of an existing column will require the entire table and its indexes to be rewritten.

    A better approach, that I believe has the same outcome if performed in a transaction is:

    ```
    ALTER TABLE "search_opinion" ADD COLUMN "joined_by_str" text;
    UPDATE "search_opinion" set "joined_by_str" = '' WHERE "joined_by_str" IS NULL;
    ALTER TABLE "search_opinion" ALTER COLUMN "joined_by_str" SET NOT NULL;
    ```

    I believe that would create a new column (instantly), set all the values to blank (slowly, but without locking or copying the table), update the column to not allow NULLs anymore, as above.

    The [django docs](https://docs.djangoproject.com/en/2.2/topics/migrations/#postgresql) also mention that:

    > The only caveat is that prior to PostgreSQL 11, adding columns with default values causes a full rewrite of the table, for a time proportional to its size. For this reason, it’s recommended you always create new columns with null=True, as this way they will be added immediately.

    So that's an option too, if you remember. I haven't figured out a way to automatically prevent this though. Maybe a git hook?

    Finally, on top of copying the whole table, after you're done, postgres doesn't reclaim the space by default. Autovacuum, as I understand it, will eventually mark the space as free, but it'll only be free to Postgres, not to the OS. To get it back in the OS, you have to run something like:

        VACUUM FULL search_opinion;

    See this bug for some details:

2. Another thing. Imagine you have two text columns you want to add. Django does this:

    ```
    --
    -- Add field correction to opinioncluster
    --
    ALTER TABLE "search_opinioncluster" ADD COLUMN "correction" text DEFAULT '' NOT NULL;
    ALTER TABLE "search_opinioncluster" ALTER COLUMN "correction" DROP DEFAULT;
    --
    -- Add field cross_reference to opinioncluster
    --
    ALTER TABLE "search_opinioncluster" ADD COLUMN "cross_reference" text DEFAULT '' NOT NULL;
    ALTER TABLE "search_opinioncluster" ALTER COLUMN "cross_reference" DROP DEFAULT;
    ```

    Instead, of running this as four (slow) queries, it could have merged them all into one (or maybe two) ALTER statements. Because each statement takes a long time, this multiplies the amount of time it takes. 

    See this bug for some details: https://code.djangoproject.com/ticket/24203

-----

There's probably more (see here: https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/), but the point is we can't trust the migration system to care about people with big systems. So we probably need to:

1. Continue generating migration files.
1. Continue generating SQL files from the migration files (`sqlmigrate`)
1. NEW: Optimize each migration SQL file before running it to fix the above.
1. Alter our migration scripts to fake the migrations instead of actually running them.

--------

I'm in the process of getting burned by this, so I'll have to revisit this when my blood is less hot and see if my solutions still sound right, but the whole thing is pretty frustrating.

[docs]: https://www.postgresql.org/docs/10/sql-altertable.html

**Comment by mlissner** [comment]
Correction: I don't think the vacuum is necessary. I just ran it and it didn't seem to make a huge difference (aside from taking a while).

**Comment by mlissner** [comment]
Another source of things you shouldn't do: https://github.com/ankane/strong_migrations#removing-a-column

**Comment by mlissner** [comment]
Another correction (nearly a year later). I don't think you can do a schema and a data migration in one transaction. The formula above probably won't work for you.

**Comment by mlissner** [comment]
We've basically stopped at this point by putting all of our notes into a MIGRATIONS.md file and being very careful. Eventually when we fix #1458, we'll do this right.
